Sources/GRPC/ReadWriteStates.swift

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2019, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import NIO
+import SwiftProtobuf
+
+/// Number of messages expected on a stream.
+enum MessageArity {
+  case none
+  case one
+  case many
+}
+
+/// Encapsulates the state required to create a new write state.
+struct PendingWriteState {
+  /// The number of messages we expect to write to the stream.
+  var arity: MessageArity
+
+  /// The encoding we should use when writing the message.
+  var encoding: CompressionMechanism
```

Comment:
Feels like we are inconsistent with encoding vs compression here; not sure if we can improve on that, though.

Reply:
The header name for compression is "grpc-encoding", and both encoding and compression are used in the spec. I'm happy to change this to `compression` here though.